### PR TITLE
Pipelines

### PR DESCRIPTION
**Pull request by viktorprogger**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | #116

### CODE REVIEW

File: src/Cli/SignalLoop.php [hunk 1]
```diff
@@ -8,8 +8,11 @@ class SignalLoop implements LoopInterface
 {
     use SoftLimitTrait;
 
+    /** @psalm-suppress UndefinedConstant */
     protected const SIGNALS_EXIT = [SIGHUP, SIGINT, SIGTERM];
```
Comment (rustamwin) [inline src/Cli/SignalLoop.php:12]:
```suggestion
    protected const SIGNALS_EXIT = [\SIGHUP, \SIGINT, \SIGTERM];
```

File: src/Cli/SignalLoop.php [hunk 2]
```diff
@@ -8,8 +8,11 @@ class SignalLoop implements LoopInterface
...
+    /** @psalm-suppress UndefinedConstant */
     protected const SIGNALS_SUSPEND = [SIGTSTP];
```
Comment (rustamwin) [inline src/Cli/SignalLoop.php:14]:
```suggestion
    protected const SIGNALS_SUSPEND = [\SIGTSTP];
```

File: src/Cli/SignalLoop.php [hunk 3]
```diff
@@ -8,8 +8,11 @@ class SignalLoop implements LoopInterface
...
+    /** @psalm-suppress UndefinedConstant */
     protected const SIGNALS_RESUME = [SIGCONT];
```
Comment (rustamwin) [inline src/Cli/SignalLoop.php:16]:
```suggestion
    protected const SIGNALS_RESUME = [\SIGCONT];
```

File: src/Middleware/InvalidCallableConfigurationException.php [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+    public function getSolution(): ?string
+    {
+        return <<<SOLUTION
+            The callable has incorrect configuration. To meet the requirements a callable should be one of:
+            - A closure.
+            - [object, method] array.
+            - [class name, method] array.
+            - [DI container service ID, method] array.
+            - DI container service ID string which references a class with the __invoke() method
```
Comment (rustamwin) [inline src/Middleware/InvalidCallableConfigurationException.php:25]:
```suggestion
            - DI container service ID string which references a class with the `__invoke()` method
```

File: README.md [hunk 5]
```diff
@@ -224,6 +207,56 @@ The component also has the ability to track the status of a job which was pushed
...
+
+You can use any of these formats to define a middleware:
+- A ready-to-use middleware object: `new FooMiddleware()`. It must implement either `MiddlewarePushInterface`,
+    or `MiddlewareConsumeInterface` depending on the place you use it.
+- An array in the format of [yiisoft/definitions](https://github.com/yiisoft/definitions).
+    **Only if you use yiisoft/definitions and yiisoft/di**.
+- A `callable`: `fn() => // do stuff`, `$object->foo(...)`, etc. It will be executed through the [yiisoft/injector](yiisoft/injector), so all the dependencies of your callable will be resolved
+- A string for your DI container to resolve the middleware, e.g. `FooMiddleware::class`
+
+The forward middleware execution will always be in the same order they are defined. I.e., if you define it like this:
```
Comment (samdark) [inline README.md:225]:
```suggestion
Middleware will be executed forwards in the same order they are defined. If you define it like the following:
```

File: README.md [hunk 6]
```diff
@@ -224,6 +207,56 @@ The component also has the ability to track the status of a job which was pushed
...
+1. `$middleware1`
+1. `$middleware2`
+1. push to a queue server *OR* handle on consume
+1. `$middleware2`
+1. `$middleware1`
+
+### Push pipeline
+When you push a message, you can use middlewares to modify both message and queue adapter. 
+With message modifying you can add extra data, obfuscate data, collect metrics, etc.  
+With queue adapter modifying you can redirect message to another queue, delay message consuming, and so on.
```
Comment (samdark) [inline README.md:236]:
```suggestion
With message modification you can add extra data, obfuscate data, collect metrics, etc.  
With queue adapter modification you can redirect message to another queue, delay message consuming, and so on.
```

File: src/Middleware/CallableFactory.php [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+     *
+     * @param mixed $definition Definition to create listener from.
+     *
+     * @throws InvalidCallableConfigurationException Failed to create listener.
+     * @throws ContainerExceptionInterface Error while retrieving the entry from container.
+     */
+    public function create(mixed $definition): callable
+    {
+        /** @var mixed */
+        $callable = $this->prepare($definition);
```
Comment (samdark) [inline src/Middleware/CallableFactory.php:39]: This can be merged into this method. By using early returns it could be simpler.

File: src/Middleware/Consume/MiddlewareFactoryConsume.php [hunk 8]
```diff
@@ -0,0 +1,96 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Yii\Queue\Middleware\Consume;
```
Comment (samdark) [inline src/Middleware/Consume/MiddlewareFactoryConsume.php:5]: What does "consume" stand for here? Do you mean "consumer"?

Comment (viktorprogger) [reply src/Middleware/Consume/MiddlewareFactoryConsume.php:5]: Consume-related middleware. The same as "push" in `Yiisoft\Yii\Queue\Middleware\Push`.

File: src/Middleware/CallableFactory.php (hunk 7 above)
Comment (viktorprogger) [reply src/Middleware/CallableFactory.php]: I've extracted the `prepare()` method. Does the result meet your expectations?

Comment (samdark) [reply src/Middleware/CallableFactory.php]: Yes.

**Review by samdark — APPROVED** [review]

File: src/Middleware/Consume/Event/AfterConsumeMiddleware.php [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace Yiisoft\Yii\Queue\Middleware\Consume\Event;
+
+use Yiisoft\Yii\Queue\Middleware\Consume\MiddlewareConsumeInterface;
+use Yiisoft\Yii\Queue\Middleware\Consume\ConsumeRequest;
+
+/**
+ * AfterMiddleware event is raised after a middleware was executed.
+ */
+final class AfterConsumeMiddleware
```
Comment (romkatsu) [inline src/Middleware/Consume/Event/AfterConsumeMiddleware.php:13]: it seems that events were recently cut out. Are we returning it again now?

File: src/Cli/SignalLoop.php [hunk 10]
```diff
@@ -8,9 +8,12 @@ class SignalLoop implements LoopInterface
 {
     use SoftLimitTrait;
 
-    protected const SIGNALS_EXIT = [SIGHUP, SIGINT, SIGTERM];
-    protected const SIGNALS_SUSPEND = [SIGTSTP];
-    protected const SIGNALS_RESUME = [SIGCONT];
+    /** @psalm-suppress UndefinedConstant */
+    protected const SIGNALS_EXIT = [\SIGHUP, \SIGINT, \SIGTERM];
```
Comment (romkatsu) [inline src/Cli/SignalLoop.php:12]: is this refactoring? Can we put it in a separate pr?

File: src/Middleware/InvalidCallableConfigurationException.php [hunk 11]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Yii\Queue\Middleware;
+
+use InvalidArgumentException;
+use Yiisoft\FriendlyException\FriendlyExceptionInterface;
+
+class InvalidCallableConfigurationException extends InvalidArgumentException implements FriendlyExceptionInterface
```
Comment (romkatsu) [inline src/Middleware/InvalidCallableConfigurationException.php:10]:
```suggestion
final class InvalidCallableConfigurationException extends InvalidArgumentException implements FriendlyExceptionInterface
```

File: src/Queue.php [hunk 12]
```diff
@@ -10,51 +10,59 @@
...
-        LoggerInterface $logger,
-        ?AdapterInterface $adapter = null,
-        string $channelName = QueueFactoryInterface::DEFAULT_CHANNEL_NAME
+        private WorkerInterface $worker,
+        private LoopInterface $loop,
+        private LoggerInterface $logger,
+        private PushMiddlewareDispatcher $pushMiddlewareDispatcher,
+        private ?AdapterInterface $adapter = null,
+        private string $channelName = QueueFactoryInterface::DEFAULT_CHANNEL_NAME,
+        MiddlewarePushInterface|callable|array|string ...$middlewareDefinitions
```
Comment (romkatsu) [inline src/Queue.php:35]: can't you do that?
```suggestion
        private MiddlewarePushInterface|callable|array|string ...$middlewareDefinitions
```

File: config/params.php [hunk 13]
```diff
@@ -14,6 +16,19 @@
     ],
     'yiisoft/yii-queue' => [
         'handlers' => [],
-        'channel-definitions' => [],
+        'channel-definitions' => [
+            'some-channel' => static fn (AdapterInterface $a): AdapterInterface => $a
+                ->withChannel('some-channel')
+                ->withPushMiddleware(SomeMiddleware::class)
```
Comment (romkatsu) [inline config/params.php:22]: there is no such class

File: src/Middleware/CallableFactory.php [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+    /**
+     * @param string $className
+     * @param string $methodName
+     *
+     * @throws ContainerExceptionInterface Error while retrieving the entry from container.
+     * @throws NotFoundExceptionInterface
+     *
+     * @return mixed
+     */
+    private function fromDefinition(string $className, string $methodName): mixed
```
Comment (romkatsu) [inline src/Middleware/CallableFactory.php:75]:
```suggestion
    private function fromDefinition(string $className, string $methodName): ?array
```

File: src/Queue.php (hunk 12 above)
Comment (viktorprogger) [reply src/Queue.php:35]: No, property promotion is not available for variadics

File: src/Middleware/CallableFactory.php (hunk 14 above)
Comment (xepozz) [reply src/Middleware/CallableFactory.php:75]: Is `?callable` possible?

File: src/Middleware/CallableFactory.php [hunk 15]
```diff
@@ -46,54 +63,44 @@ public function create(mixed $definition): callable
...
+     *
+     * @throws ContainerExceptionInterface Error while retrieving the entry from container.
+     * @throws NotFoundExceptionInterface
      */
-    private function prepare(mixed $definition): mixed
+    private function fromDefinition(string $className, string $methodName): mixed
     {
-        if (is_string($definition) && $this->container->has($definition)) {
-            return $this->container->get($definition);
+        if (!class_exists($className) && $this->container->has($className)) {
```
Comment (xepozz) [inline src/Middleware/CallableFactory.php:77]: Can the first check be omitted and the further fetching from the container (on 97 line) also be removed?

File: README.md [hunk 16]
```diff
@@ -224,6 +207,56 @@ The component also has the ability to track the status of a job which was pushed
...
+- A `callable`: `fn() => // do stuff`, `$object->foo(...)`, etc. It will be executed through the [yiisoft/injector](yiisoft/injector), so all the dependencies of your callable will be resolved
+- A string for your DI container to resolve the middleware, e.g. `FooMiddleware::class`
+
+Middleware will be executed forwards in the same order they are defined. If you define it like the following:
+`[$middleware1, $midleware2]`, the execution will look like this:
+1. `$middleware1`
+1. `$middleware2`
+1. push to a queue server *OR* handle on consume
+1. `$middleware2`
+1. `$middleware1`
```
Comment (samdark) [inline README.md:231]:
```suggestion
```mermaid
graph LR
    StartPush((Start)) --> PushMiddleware1[$middleware1] --> PushMiddleware2[$middleware2] --> Push(Push to a queue)
    -.-> PushMiddleware2[$middleware2] -.-> PushMiddleware1[$middleware1]
    PushMiddleware1[$middleware1] -.-> EndPush((End))
    

    StartConsume((Start)) --> ConsumeMiddleware1[$middleware1] --> ConsumeMiddleware2[$middleware2] --> Consume(Consume / handle)
    -.-> ConsumeMiddleware2[$middleware2] -.-> ConsumeMiddleware1[$middleware1]
    ConsumeMiddleware1[$middleware1] -.-> EndConsume((End))
```
```

Comment (samdark) [reply README.md:231]: That would be:

```mermaid
graph LR
    StartPush((Start)) --> PushMiddleware1[$middleware1] --> PushMiddleware2[$middleware2] --> Push(Push to a queue)
    -.-> PushMiddleware2[$middleware2] -.-> PushMiddleware1[$middleware1]
    PushMiddleware1[$middleware1] -.-> EndPush((End))
    
    StartConsume((Start)) --> ConsumeMiddleware1[$middleware1] --> ConsumeMiddleware2[$middleware2] --> Consume(Consume / handle)
    -.-> ConsumeMiddleware2[$middleware2] -.-> ConsumeMiddleware1[$middleware1]
    ConsumeMiddleware1[$middleware1] -.-> EndConsume((End))
```

**Review by samdark — APPROVED** [review]

File: src/Middleware/InvalidCallableConfigurationException.php [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+namespace Yiisoft\Yii\Queue\Middleware;
+
+use InvalidArgumentException;
+use Yiisoft\FriendlyException\FriendlyExceptionInterface;
+
+final class InvalidCallableConfigurationException extends InvalidArgumentException implements FriendlyExceptionInterface
+{
+    public function getName(): string
+    {
+        return 'Invalid event listener configuration.';
```
Comment (romkatsu) [inline src/Middleware/InvalidCallableConfigurationException.php:14]: Is the phrase "listener" exactly appropriate here?

File: src/Worker/Worker.php [hunk 18]
```diff
@@ -41,7 +41,7 @@ public function __construct(
      *
      * @throws Throwable
      */
-    public function process(MessageInterface $message, QueueInterface $queue): void
+    public function process(MessageInterface $message, QueueInterface $queue): MessageInterface
```
Comment (romkatsu) [inline src/Worker/Worker.php:44]: PHPDoc comment signature is not complete

File: src/Middleware/Consume/MiddlewareFactoryConsumeInterface.php [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+declare(strict_types=1);
+
+namespace Yiisoft\Yii\Queue\Middleware\Consume;
+
+/**
+ * Creates a middleware based on the definition provided.
+ * You may implement this interface if you want to introduce custom definitions or pass additional data to
+ * the middleware created.
+ */
+interface MiddlewareFactoryConsumeInterface
```
[the remark this answers is not in the document]
Comment (viktorprogger) [reply src/Middleware/Consume/MiddlewareFactoryConsumeInterface.php:12]: Middlewares here are a copy of `yiisoft/middleware-dispatcher` ones. I think, the cases are the same. Some people will use another container with another form of definitions. So, they won't want to write yiisoft-specific definitions for queues only.
